Unified heap allocation interface

deyaaeldeen opened the pull request:
The key contribution is the `define-memory-layout-helpers` macro that manages how heap-allocated objects are created, accessed, and checked for equality. I used it for allocating all our types so far. I will use it for coercions too but after I finish the runtime hashconsing for types first.

akuhlens commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
+                      [(sindex-def* ...) index-def*]
+                      [(sindex-data-def* ...) index-data-def*]
+                      [tag-data-def (gen-data-id #'tag-def)]
+                      [(sindex-val* ...)
```
This is the same as `(datum->syntax stx (range ...))`

deyaaeldeen replied:
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
-            `(("type" . ,(compile-type-static/dynamic ty)))))
+(define-syntax (memory-layout$ stx)
+  (syntax-case stx ()
+    [(_ namespace name tag (field* cval* mode*) ...)
```
I agree, it is for statically known values. It will be used for things like tagging med coercions.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
+                 (values (cons (Quote cdtm) p*) t*)
+                 (with-syntax ([a (generate-temporary)])
+                   (values (cons #'a p*) (cons #'a t*)))))))
+       (define (all-but-last l)
```
Thanks for pointing this out, I should have recognized that.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
+                          (map gen-func-access
+                               field-string* mode-symbol*
+                               index-def* index-data-def*)])
+             #'(begin
```
I get this error `unsyntax-splicing: illegal outside of quasisyntax`.

akuhlens replied:
Did you notice the 
```
#`
```
instead of
```
 #'
```
?

deyaaeldeen replied:
you're right, done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -128,7 +131,7 @@ but a static single assignment is implicitly maintained.
                  [tag-shifted-imm
                   (lambda ([tag : D0-Expr]) : D0-Expr
-                    (Op 'binary-or `(,shifted-imm ,tag)))])
+                          (Op 'binary-or `(,shifted-imm ,tag)))])
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
-  (sr-alloc "MVectT" TYPE-MVECT-TAG
-            `(("type" . ,(compile-type-static/dynamic ty)))))
+(define-syntax (memory-layout$ stx)
+  (syntax-case stx ()
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
+         (cond [(empty? l) (error 'all-but-last "empty list")]
+               [(empty? (rest l)) empty]
+               [else (cons (first l) (all-but-last (rest l)))]))
+       (with-syntax* ([func-alloc-name
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
+                               (sr-alloc sname-string tag-def
+                                         `((sfield-string* . ,alloc-val*) ...)))
+                           (with-syntax ([arg (generate-temporary)]
+                                         [(alloc-arg*x ...) (all-but-last temp*)]
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
+                                            (cons last-field-string arg))
+                                          last-alloc-arg))))))]
+                      [equal-arg (generate-temporary)])
+         (letrec ([gen-func-access
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1442,132 @@ but a static single assignment is implicitly maintained.
+                   (sr-check-tag=? equal-arg namespace-mask-def tag-def))
+                 func-access ...)))))]))
+
+(memory-layout$ type gref #b001 ("type" #f single))
```
I changed all of them to strings.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -148,7 +151,10 @@ but a static single assignment is implicitly maintained.
-                     (assign$ t (Op 'Alloc (list (sr-plus count (Quote 1)))))
+                     (assign$
+                      t
+                      (Op 'Alloc (list (sr-plus count
```
I added a new commit to fix these.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1031,9 +1021,9 @@ but a static single assignment is implicitly maintained.
-          (list (Op 'Print (list e))
-                (Op 'Exit  (list (Quote -1))))
-          UNDEF-IMDT)]
+           (list (Op 'Print (list e))
```
Done.

deyaaeldeen commented:
@akuhlens I addressed all your comments, swapped the order of two commits and added two more commits.

akuhlens commented:
Getting really close.

Tons of places where the indentation could be fixed, many are commented but some are not.
Here is the list of indentations schemes I am using for racket mode which should mostly automate
the process. 
```scheme
'((let$* 1)                                                             
  (let*$ 1)                                                             
  (let$  1)                                                             
  (Let 1)                                                               
  (Lambda 1)                                                            
  (Letrec 1)                                                            
  (Code 1)                                                              
  (code$ 1)                                                             
  (case$ 1)                                                             
  (Switch 1)                                                            
  (Letrec 1)                                                            
  (Repeat 5)                                                            
  (repeat$ 2)                                                           
  (precondition$ 1)                                                     
  (Prog 1)                                                              
  (GlobDecs 1)                                                          
  (Labels 1))
```
This file in particular might be helped by introducing `assign$` and `op$` with indentation setting of 1. This would effectively give assign$ a define like indentation and op$ would treat its first argument as part of the operator so far as the indentation is concerned.

If there is some technology that would allow us to add this to the repo I would be interested in making the sharing of indentations more automated.

If you care to make `define-memory-layout-helpers` cleaner, I think the pattern could be more specific and based off this change several more changes will be available.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1746,13 +1683,13 @@ but a static single assignment is implicitly maintained.
-      (if (uid=? c clos)
-          (hash-ref map f (fvar-err f))
-          (clos-err c f)))))
+            (if (uid=? c clos)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1246,60 +1154,49 @@ but a static single assignment is implicitly maintained.
-           (assign$ new-tpl (Op 'Alloc (list n)))
+           (assign$ new-tpl (op$ Alloc n))
            (repeat$ (i ZERO-IMDT n) (_ UNIT-IMDT)
                     (begin$
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1116,87 +1036,77 @@ but a static single assignment is implicitly maintained.
-         (define arg : D0-Expr
-           (Op 'binary-xor (list e TYPE-MVECT-TAG)))
-         (Op 'Array-ref (list arg TYPE-MVECT-TYPE-INDEX))]
+               (If (op$ >= ind ZERO-IMDT) ;; vectors indices starts from 0
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1026,83 +952,77 @@ but a static single assignment is implicitly maintained.
+           (op$ Array-set! vect UGVECT-SIZE-INDEX size)
            (repeat$ (i UGVECT-OFFSET repr-size) (_ UNIT-IMDT)
-                    (Op 'Array-set! (list vect i init-val)))
+                    (op$ Array-set! vect i init-val))
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -983,39 +913,35 @@ but a static single assignment is implicitly maintained.
            (Switch tag
-                   `([(,data:DYN-BOXED-TAG) . ,(Op 'Array-ref
-                                                   (list tmp DYN-TYPE-INDEX))]
+                   `([(,data:DYN-BOXED-TAG) . ,(op$ Array-ref tmp DYN-TYPE-INDEX)]
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -595,15 +564,15 @@ but a static single assignment is implicitly maintained.
       (define comp-tuple-crcn-c : D0-Code
         (code$ (crcn1 crcn2 i count id?1)
-               (If (Op '= `(,i ,count))
+               (If (op$ = i count)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -502,22 +479,18 @@ but a static single assignment is implicitly maintained.
         (code$ (tpl-val t1 t2 l i count)
-               (If (Op '= `(,i ,count))
-                   (begin$ (assign$ crcn (Op 'Alloc (list count))) crcn)
+               (If (op$ = i count)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -416,26 +398,24 @@ but a static single assignment is implicitly maintained.
+                  (op$ %>> tagged-count COERCION-SECOND-TAG-SHIFT))
                  (repeat$ (i ZERO-IMDT count) (_ UNIT-IMDT)
-                          (assign$ val (Op 'Array-ref (list tpl-val i)))
+                          (assign$ val (op$ Array-ref tpl-val i))
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -375,10 +357,10 @@ but a static single assignment is implicitly maintained.
-               (If (Op '= `(,i ,count))
-                   (begin$ (assign$ tpl (Op 'Alloc (list count))) tpl)
+               (If (op$ = i count)
+                   (begin$ (assign$ tpl (op$ Alloc count)) tpl)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -375,10 +357,10 @@ but a static single assignment is implicitly maintained.
         (code$ (v c i count)
-               (If (Op '= `(,i ,count))
-                   (begin$ (assign$ tpl (Op 'Alloc (list count))) tpl)
+               (If (op$ = i count)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -294,7 +280,7 @@ but a static single assignment is implicitly maintained.
       (define comp-fn-c : D0-Code
         (code$ (c1 c2 i arity id?1)
-               (If (Op '= `(,i ,arity))
+               (If (op$ = i arity)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -240,7 +227,7 @@ but a static single assignment is implicitly maintained.
       (define mk-fn-crcn-c : D0-Code
         (code$ (t1 t2 l i arity)
-               (If (Op '= `(,i ,arity))
+               (If (op$ = i arity)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -191,7 +181,7 @@ but a static single assignment is implicitly maintained.
       (define mk-fn-type-glb-c : D0-Code
         (code$ (t1 t2 i arity)
-               (If (Op '= `(,i ,arity))
+               (If (op$ = i arity)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -146,24 +139,21 @@ but a static single assignment is implicitly maintained.
       (define mk-tuple-type-glb-c : D0-Code
         (code$ (t1 t2 i count)
-               (If (Op '= `(,i ,count))
+               (If (op$ = i count)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1352,144 @@ but a static single assignment is implicitly maintained.
+       (define/with-syntax name
+         (format-id stx "~a-~a-~a-access" namespace-string name-string x))
+       (define/with-syntax access-arg (generate-temporary))
+       (if (eq? y 'single)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1352,144 @@ but a static single assignment is implicitly maintained.
+     (define/with-syntax (alloc-arg* ...) temp*)
+     (define/with-syntax (alloc-val* ...) field-val*)
+     (define/with-syntax func-alloc
+       (if (eq? (last mode-symbol*) 'single)
```
Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1352,144 @@ but a static single assignment is implicitly maintained.
+    #:datum-literals (single many)
+    [(_ namespace:str name:str tag:exact-integer
+        (field*:str (~optional cval* #:defaults ([cval* #'#f]))
+                    (~and mode*:id (~or single many))) ...+)
```
It is not quite that but I got the idea. Some of the code became clearer but some became a little awkward too. Done.

deyaaeldeen commented on src/casts/specify-representation.rkt:
```diff
@@ -1463,54 +1352,144 @@ but a static single assignment is implicitly maintained.
-(define (sr-type-mvect compile-type-static/dynamic ty)
-  (sr-alloc "MVectT" TYPE-MVECT-TAG
-            `(("type" . ,(compile-type-static/dynamic ty)))))
+;; it generates constants and functions for creating, accessing, and checking
```
Done.

deyaaeldeen commented:
@akuhlens I addressed all your comments. However having `assign$` of indentation level 1 produces ugly indentation, for instance:
```
(assign$
     t1a
  (sr-tagged-array-ref
   t1 TYPE-TUPLE-TAG (sr-plus TYPE-TUPLE-ELEMENTS-OFFSET i)))
```

I changed the order of the commits again so that the macro commit is on top and I made it cleaner.

akuhlens commented:
Indentaion 1 indicates that it will indent like let. For every assign$ that needs more than one line should be indented like:
```racket
(assign$ t1a 
   (sr-tagged-array-ref 
       t1 TYPE-TUPLE-TAG (sr-plus TYPE-TUPLE-ELEMENTS-OFFSET i)))
```

deyaaeldeen commented:
@akuhlens gotcha, done.
